src/front/glsl/preprocessor.rs

```diff
@@ -0,0 +1,1730 @@
+use super::{
+    super::{Error, ErrorKind},
+    lex, Literal, Token, TokenMetadata,
+};
+use crate::FastHashMap;
+use std::{
+    fmt,
+    iter::Peekable,
+    ops::{Deref, Range},
+    vec::IntoIter,
+};
+
+#[derive(Debug)]
+pub enum PreprocessorIfNode {
+    Literal(Literal),
+    Unary {
+        op: UnaryOp,
+        tgt: Box<PreprocessorIfNode>,
+    },
+    Binary {
+        left: Box<PreprocessorIfNode>,
+        op: BinaryOp,
+        right: Box<PreprocessorIfNode>,
+    },
+}
+
+impl Literal {
+    pub fn as_isize(&self) -> isize {
+        match self {
+            Literal::Double(double) => *double as isize,
+            Literal::Float(float) => *float as isize,
+            Literal::Uint(uint) => *uint as isize,
+            Literal::Sint(sint) => *sint,
+            Literal::Bool(val) => *val as isize,
+        }
+    }
+
+    pub fn as_bool(&self) -> Result<bool, Error> {
+        Ok(match self {
+            Literal::Double(_) | Literal::Float(_) => panic!(),
+            Literal::Uint(uint) => {
+                if *uint == 0 {
+                    false
+                } else {
+                    true
+                }
+            }
+            Literal::Sint(sint) => {
+                if *sint == 0 {
+                    false
+                } else {
+                    true
+                }
+            }
+            Literal::Bool(val) => *val,
+        })
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum UnaryOp {
+    Positive,
+    Negative,
+    BitWiseNot,
+    LogicalNot,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum BinaryOp {

```

Comment:
do we really need to have a separate enum for this, given that `crate::BinaryOperator` exists?